### PR TITLE
pluginlink: hooking and calling helpers that exit(1) on failure

### PR DESCRIPTION
**Pull request by niv**
This is mostly for lazy plugin authors (like me) that want to hook
events or call services and just fail when they don't exist, because
the plugin would not work otherwise.

Best used in OnPluginsLoaded or similar - fail early and hard, like you
read about.

--

What do you think?

(In any case don't merge yet, if this is desirable I'll add optional/better error messages and warnings to the default macros too).

### CODE REVIEW

**Comment by virusman** [comment]
That can be useful. Not sure about CallServiceOrFail, though: service functions are usually called at runtime.

**Comment by niv** [comment]
Alternative suggestion: I'd even go further and make the abort() the default, and add an optional param or a different macro to make it non-aborting for the one or two corner cases where a failed hook is something the plugin knows to handle.

How about that?

**Comment by niv** [comment]
Branch reworked to make the default impl fail, instead of having a separate macro.

Also adds ```HookOptionalEvent()``` macro for the (very few) cases where event hooks aren't mandatory.

Pulls in <execinfo.h>, which is needed for backtraces! Shouldn't be a problem though.

Note that this is still just a request-for-comment, hasn't touched all the in-tree plugins yet!

**Comment by niv** [comment]
Branch updated again! This time with less breakage. Please see the commit message for details.

tldr: HookEvent() now fails hard, HookEventOptionally() for those that want to check themselves. ABI change to pluginLink by adding a new function at the end.

A sample failure case would look like this (tested with a outdated nwnx_optimizations):

```
[PluginLink] a plugin requires hook 'Core/MainLoopBefore', which noone provides.
Backtrace:
 - ./nwnx2.so(HookEvent+0x68) [0xb7756baa]
 - ./nwnx_optimizations.so(_Z13HookFunctionsv+0x1d) [0xb400672b]
 - ./nwnx_optimizations.so(_ZN18CNWNXOptimizations8OnCreateEP5glinePKc+0x80) [0xb400694e]
 - ./nwnx2.so(_Z13LoadLibrariesv+0x469) [0xb7752cf6]
 - ./nwnx2.so(_ZN9startstopC2Ev+0x18e) [0xb77538ce]
 - ./nwnx2.so(+0xe19d5) [0xb77539d5]
 - ./nwnx2.so(+0xe1a09) [0xb7753a09]
 - /lib/ld-linux.so.2(+0xe82e) [0xb77da82e]
 - /lib/ld-linux.so.2(+0xe924) [0xb77da924]
 - /lib/ld-linux.so.2(+0xd3f) [0xb77ccd3f]
Aborted (core dumped)
```

**Comment by virusman** [comment]
Looks good to me.

**Comment by niv** [comment]
So, merge then? :)

**Comment by virusman** [comment]
Thanks!
